File: 백준/Silver/1497. 기타콘서트/기타콘서트.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

    static int n, m;
    static long maxPossible;
    static int ans = -1;
    static long diff = Long.MAX_VALUE;
    static long[] arr;

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());

        arr = new long[n];

        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());

            String name = st.nextToken();
            String song = st.nextToken();

            for (int j = 0; j < m; j++) {
                if (song.charAt(j) == 'Y') {
                    arr[i] |= (1L << (m - j - 1));
                }
            }
        }

        maxPossible = (1L << m) - 1;

        for (int i = 1; i <= n; i++) {
            solve(0, i, 0, 0);
        }

        System.out.println(ans);
    }

    private static void solve(int depth, int limit, int visit, int start) {

        if (depth == limit) {
            check(visit, limit);
            return;
        }

        for (int i = start; i < n; i++) {

            if ((visit & (1 << i)) != 0) {
                continue;
            }

            solve(depth + 1, limit, visit | (1 << i), i + 1);
        }
    }

    private static void check(int visit, int limit) {

        long temp = 0L;

        for (int i = 0; i < n; i++) {

            if ((visit & (1 << i)) != 0) {

                temp |= arr[i];
            }
        }

        if (temp != 0) {
            long d = maxPossible - temp;
            if (d < diff) {
                diff = d;
                ans=Math.max(ans, limit);
            }
        }
    }
}

File: 백준/Silver/1497. 기타콘서트/기타콘서트.py
n, m = map(int, input().split())
max_possible = 0
min_used = float('inf')

arr = [0] * n

for i in range(n):
    name, song = input().split()

    for j in range(m):
        if song[j] == 'Y':
            arr[i] |= (1 << (m - j - 1))


def solve(depth, possible, used):
    global max_possible, min_used

    possible_count = bin(possible).count('1')

    if possible_count == max_possible:
        min_used = min(min_used, used)
    elif possible_count > max_possible:
        max_possible = possible_count
        min_used = used

    if depth == n:
        return

    solve(depth + 1, possible, used)
    solve(depth + 1, possible | arr[depth], used + 1)


solve(0, 0, 0)

print(min_used if not max_possible == 0 else -1)
